test/parse.jl

```diff
@@ -1195,3 +1195,26 @@ end
 # issue #16937
 @test expand(:(f(2, a=1, w=3, c=3, w=4, b=2))) == Expr(:error,
                                                        "keyword argument \"w\" repeated in call to \"f\"")
+
+# Basic parsing of reference syntax
+@test expand(parse("a@b")) == :(gepfield(a, :b))
+@test expand(parse("a@c")) == :(gepfield(a, :c))
+@test expand(parse("a@c.d")) == :(gepfield(gepfield(a, :c), :d))
+@test expand(parse("a@c.d[e]")) == :(gepindex(gepfield(gepfield(a, :c), :d), e))
+@test expand(parse("a@c[d]")) == :(gepindex(gepfield(a, :c), d))
+@test expand(parse("a@c[d,e]")) == :(gepindex(gepfield(a, :c), d, e))
+@test expand(parse("a@[c]")) == :(gepindex(a, c))
+@test expand(parse("a@[c,d]")) == :(gepindex(a, c, d))
+@test expand(parse("a@[c].d")) == :(gepfield(gepindex(a, c),:d))
+
+@test expand(parse("a@b = 1")) == :(setindex!(gepfield(a, :b), 1))
+@test expand(parse("a@c = 1")) == :(setindex!(gepfield(a, :c), 1))
+@test expand(parse("a@c.d = 1")) == :(setindex!(gepfield(gepfield(a, :c), :d), 1))
+@test expand(parse("a@c.d[e] = 1")) == :(setindex!(gepindex(gepfield(gepfield(a, :c), :d), e), 1))
+@test expand(parse("a@c[d] = 1")) == :(setindex!(gepindex(gepfield(a, :c), d), 1))
+@test expand(parse("a@c[d,e] = 1")) == :(setindex!(gepindex(gepfield(a, :c), d, e), 1))
+@test expand(parse("a@[c] = 1")) == :(setindex!(gepindex(a, c), 1))
+@test expand(parse("a@[c,d] = 1")) == :(setindex!(gepindex(a, c, d), 1))
+@test expand(parse("a@[c].d = 1")) == :(setindex!(gepfield(gepindex(a, c),:d), 1))
+
+expand(:(foo(1)@a))
```

Comment:
shouldn't crash, the LHS of an `@` can in principle be any expression. Writing the expression it expands to is a bit complicated though (has a bunch of SSAVals). I'll probably remove that here for the final version in favor of actually testing that it does what you expect semantically.